contributors/design-proposals/node/numa-manager.md

```diff
@@ -0,0 +1,281 @@
+this situation needs more discussion. For example, the NUMA manager
+could enforce strict NUMA alignment for Guaranteed QoS pods.
+Alternatively, the NUMA manager could simply provide best-effort NUMA
+alignment for all pods.
+
+The NUMA Manager component will be disabled behind a feature gate until
+graduation from alpha to beta.
+
+#### Computing Preferred Affinity
+
+A NUMA hint is a list of possible NUMA node masks. After collecting hints
+from all providers, the NUMA Manager must choose some mask that is
+present in all lists. Here is a sketch:
+
+1. Apply a partial order on each list: number of bits set in the
+   mask, ascending. This biases the result to be more precise if
+   possible.
+1. Iterate over the permutations of preference lists and compute
+   bitwise-and over the masks in each permutation.
+1. Store the first non-empty result and break out early.
+1. If no non-empty result exists, return an error.
+
+#### New Interfaces
+
+```go
+package numamanager
+
+// NUMAManager helps to coordinate NUMA-related resource assignments
+// within the Kubelet.
+type Manager interface {
+  lifecycle.PodAdmitHandler
+  Store
+  AddHintProvider(HintProvider)
+  RemovePod(podName string)
+}
+
+// NUMAMask is a bitmask-like type denoting a subset of available NUMA nodes.
+type NUMAMask struct{} // TBD
+
+// NUMAStore manages state related to the NUMA manager.
+type Store interface {
+  // GetAffinity returns the preferred NUMA affinity for the supplied
+  // pod and container.
+  GetAffinity(podName string, containerName string) NUMAMask
+}
+
+// HintProvider is implemented by Kubelet components that make
+// NUMA-related resource assignments. The NUMA manager consults each
+// hint provider at pod admission time.
+type HintProvider interface {
+  GetNUMAHints(pod v1.Pod, containerName string) []NUMAMask
+}
+```
+
+_NUMA Manager and related interfaces (sketch)._
+
+![numa-manager-components](https://user-images.githubusercontent.com/379372/35370509-13dd9488-0143-11e8-998b-6b5115982842.png)
+
+_NUMA Manager components._
+
+![numa-manager-instantiation](https://user-images.githubusercontent.com/379372/35370513-17f90f70-0143-11e8-88e3-f199e9717946.png)
+
+_NUMA Manager instantiation and inclusion in pod admit lifecycle._
+
+### Changes to Existing Components
+
+1. Kubelet consults NUMA Manager for pod admission (discussed above.)
+1. Add two implementations of NUMA Manager interface and a feature gate.
+    1. As much NUMA Manager functionality as possible is stubbed when the
+       feature gate is disabled.
+    1. Add a functional NUMA manager that queries hint providers in order
+       to compute a preferred NUMA node mask for each container.
+1. Add `GetNUMAHints()` method to CPU Manager.
+    1. CPU Manager static policy calls `GetAffinity()` method of NUMA
+       manager when deciding CPU affinity.
+1. Add `GetNUMAHints()` method to Device Manager.
+    1. Add NUMA Node ID to Device structure in the device plugin
```

Comment:
We might not want to add NUMA ID specifically in the device structure, instead use key-value attributes map:
https://github.com/kubernetes/kubernetes/pull/58388

Reply:
Thanks for linking that PR. My preference would be to avoid using semi structured K-V for any values the Kubelet cares about. We can get some more opinions on that. Either way, fairly minor detail.

Reply:
I am also not sure adding specific fields in DP, +1 for some general types (but not raw K-V, painful for programming ... )

Reply:
@kad this is interesting observation because I sort of came to this conclusion working through how I would partition a system for dejittering.

The way RH does this is through the use of a tuned profile called "cpu-partitioning".  It also uses cpu sets.

Reply:
Would you all be OK with the following approach?

1. Generalize the hint structure from `NUMAHint` to something like `TopologyHints`.
1. `TopologyHints` structure has a field for `SocketID`

The material in this proposal, and known use cases, are really more about alignment with a physical socket and e.g. it's local PCI complex than "NUMA". That way, if we find use cases for specifically talking about memory nodes, or sub-socket CPU topologies, we can extend the hint structure.

We may even want to rename the component from "NUMA Manager" to something more generic like "Topology Manager" accordingly.

@kad @jeremyeder

Reply:
Agreed, since the eventual intent is to handle things like PCI locality on behalf of NVLINK or NICs, and that's only related to NUMA itself in the context of Intel's current chip design.

Reply:
"Topology Manager" sounds to me a lot better than "NUMA Manager".
Generalising topology hints as a struct makes sense. However, `SocketID` doesn't really make sense.
As a customer of that API in device plugin, for me it is hard to get which `SocketID` it will be. It will be a lot easier to read corresponding to device sysfs `local_cpulist` entry (e.g. `/sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0/local_cpulist`) and pass it up to `Topology Manager`.
If `Topology Manager` policies will be interested in some kind of grouping of those, it can be done outside of plugins.

Making hints structure of containing two hints fields (cpu set and memory set, similar like it is defined in 
https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#cpu for fields `cpu.cpus` and `cpu.mems`) will allow to cover universally any topology for both IO critical or Memory critical alignments.

Reply:
The points of mistakenly referring to a socket as a single NUMA node, and also the likely need for future kinds of locality hints are well-taken. Thanks for bringing those up.

In the interest of maintaining momentum on this proposal and leaving the door open to further hints, I've changed the naming and suggested API to a struct of hints as described above. We can explore more granular hints and related semantics in subsequent proposals. This is a deep topic. The semantics of more granularity are going to require some extended discussions, and currently the Kubelet does not set `cpuset.mems` at all, so it may be a stretch to get that (unenforceable) piece into the API at this point. At the same time, most participants in the discussion have been supportive / in agreement about the goal of at least socket-level alignment. We discussed using the CPU as a reference for internal node topology way back during the first resource management wg face-to-face in early 2017. Hopefully everyone is OK with this compromise for the first phase.